### ISSUE TITLE
Bug using nested "with" statements

### ISSUE THREAD

**Issue by inlinevoid** [leafo/moonscript#220]
Not sure how else to describe it, but here's some sample code:
```MoonScript
class Grid
	child:		nil
	on_draw:	nil

	draw: =>
		@\on_draw! unless @on_draw == nil
		@child\draw! unless @child == nil


MainWindow = with Grid!
	.child = with Grid!
		.on_draw = () =>
			consolePrint "foo"
```

And the resulting Lua (comment near the bottom highlighting the problem):
```Lua
local Grid
do
  local _base_0 = {
    child = nil,
    on_draw = nil,
    draw = function(self)
      if not (self.on_draw == nil) then
        self:on_draw()
      end
      if not (self.child == nil) then
        return self.child:draw()
      end
    end
  }
  _base_0.__index = _base_0
  local _class_0 = setmetatable({
    __init = function() end,
    __base = _base_0,
    __name = "Grid"
  }, {
    __index = _base_0,
    __call = function(cls, ...)
      local _self_0 = setmetatable({}, _base_0)
      cls.__init(_self_0, ...)
      return _self_0
    end
  })
  _base_0.__class = _class_0
  Grid = _class_0
end
local MainWindow
do
  local _with_0 = Grid()
  do
    local _with_1 = Grid()
    _with_1.on_draw = function(self)
      return consolePrint("foo")
    end
    _with_1.child = _with_1 -- This should actually be _with_0.child = _with_1
  end
  MainWindow = _with_0
end
```

Really enjoying MoonScript btw.  It's made working with Lua way more fun.

**Comment by kirbyfan64** [comment]
Is this a duplicate of #110?

**Comment by kirbyfan64** [comment]
And #197. This was fixed in master recently (I *knew* I'd seen it before! :).

**Comment by inlinevoid** [comment]
I think you're right.  Just noticed the build I'm using, 0.3.0, is 9 months old.  Don't have the time to learn how to compile master right now so I'll report back in a day or so.

**Comment by kirbyfan64** [comment]
Just run `make`!

**Comment by leafo** [comment]
Luarocks also has the latest version with the fix: https://luarocks.org/modules/leafo/moonscript

**Comment by inlinevoid** [comment]
Alright so I just built the same sample as above with 0.3.2.  It still produces incorrect code

```Lua
local Grid
do
  local _class_0
  local _base_0 = {
    child = nil,
    on_draw = nil,
    draw = function(self)
      if not (self.on_draw == nil) then
        self:on_draw()
      end
      if not (self.child == nil) then
        return self.child:draw()
      end
    end
  }
  _base_0.__index = _base_0
  _class_0 = setmetatable({
    __init = function() end,
    __base = _base_0,
    __name = "Grid"
  }, {
    __index = _base_0,
    __call = function(cls, ...)
      local _self_0 = setmetatable({}, _base_0)
      cls.__init(_self_0, ...)
      return _self_0
    end
  })
  _base_0.__class = _class_0
  Grid = _class_0
end
local MainWindow
do
  local _with_0 = Grid()
  do
    local _with_1 = Grid()
    _with_1.on_draw = function(self)
      return consolePrint("foo")
    end
    _with_1.child = _with_1 -- Should be _with_0.child = _with_1
  end
  MainWindow = _with_0
end
```

**Comment by kirbyfan64** [comment]
@leafo But wasn't the bugfix merged in a month ago, and the latest LuaRocks release is older?

**Comment by inlinevoid** [comment]
I didn't use LuaRocks, I compiled from master.

**Comment by inlinevoid** [comment]
Actually, I did use LuaRocks to build and gather dependencies but the MoonScript source is from master.

**Comment by leafo** [comment]
ah, woops. Okay I'll release a new version.

**Comment by leafo** [comment]
0.4.0 is out now

**Comment by Nymphium** [comment]
It is NOT fixed, and I get a similar bug.

```moonscript
k = with {}
  .foo = with str = ""
    str ..= "x"
```

and be compiled like:

```lua
local k
do
  local _with_0 = { }
  do
    local str = ""
    str = str .. "x"
    str.foo = str  -- Oh, No! expected `k.foo = str`
  end
  k = _with_0
end
```

Yes, this code has an error with the HEAD version:

```
-- expected: k.foo = "3"
--   actual: 
-- moon: /home/nymphium/works/test/bug.moon:2: (7) attempt to index a string value (local 'str')
```

**Comment by brashaw** [comment]
@leafo, @Nymphium is correct.  This is not fixed in [version 0.4.0](https://github.com/leafo/moonscript/releases/tag/v0.4.0).

```Lua
local Grid
do
  local _class_0
  local _base_0 = {
    child = nil,
    on_draw = nil,
    draw = function(self)
      if not (self.on_draw == nil) then
        self:on_draw()
      end
      if not (self.child == nil) then
        return self.child:draw()
      end
    end
  }
  _base_0.__index = _base_0
  _class_0 = setmetatable({
    __init = function() end,
    __base = _base_0,
    __name = "Grid"
  }, {
    __index = _base_0,
    __call = function(cls, ...)
      local _self_0 = setmetatable({}, _base_0)
      cls.__init(_self_0, ...)
      return _self_0
    end
  })
  _base_0.__class = _class_0
  Grid = _class_0
end
local MainWindow
do
  local _with_0 = Grid()
  do
    local _with_1 = Grid()
    _with_1.on_draw = function(self)
      return consolePrint("foo")
    end
    _with_1.child = _with_1
  end
  MainWindow = _with_0
end
```

This is the same output generated by version 0.3.2.
